File: Preprocessing/CrystalGeneratrion/CreateElectronDist.py
import numpy as np
import os
import matplotlib.pyplot as plt
import math
import cmath

#print u'\u212B'.encode('utf-8')
print u'你'.encode('utf-8')

Path = "/home/jallpa/Documents/GitHub/FelixML_Project/Preprocessing/NextPart"
StructFact = "/StructureFactor.txt"

PathOfFile = Path + StructFact
with open(PathOfFile) as textFile:
    lines = [line.split() for line in textFile]

FileLength = len(lines)

gVectors = np.zeros(FileLength * 3).reshape(FileLength, 3)
Ug = np.zeros(FileLength * 2, dtype = np.cdouble).reshape(FileLength, 2)
for i in range(0, FileLength):
    gVectors[i][0] = int(lines[i][0])
    gVectors[i][1] = int(lines[i][1])
    gVectors[i][2] = int(lines[i][2])
    Ug[i][0] = float(lines[i][5])
    Ug[i][1] = float(lines[i][6])

#print(Ug)

nPix = 128

Vx = np.array([1, 0, 0])
Vy = np.array([0, 1, 0])
MagX = 1
MagY = 1

x = np.arange(nPix, dtype = np.float) / (nPix - 1)
y = np.arange(nPix, dtype = np.float) / (nPix - 1)

RScattFacToVolts = 47.913838
constPi = 2 * 3.1415926535 * 1j

RreUr = np.zeros(nPix * nPix).reshape(nPix, nPix)
RimUr = np.zeros(nPix * nPix).reshape(nPix, nPix)

Re = [0, 0, 0]
for i in range(0, nPix):
    for j in range(0, nPix):
        Re = x[i] * Vx + y[j] * Vy
        #print(Re)
        for n in range(0, FileLength):
            cug = Ug[n][0] + 1j * Ug[n][1]

            ReDotgVec = np.dot(Re, gVectors[n])

            ExpVar = cmath.exp(constPi * ReDotgVec)
            #print(ReDotgVec)
            RreUr[j][i]+= (cug * ExpVar).real
            #RimUr[j][i]+=  (1j * cug * ExpVar).imag



RreUr = RreUr * RScattFacToVolts
#RimUr = RimUr * RScattFacToVolts
print(RreUr)


fig=plt.figure(figsize = (32, 32))
img = RreUr
plt.imshow(img, cmap = "gray")
plt.legend()
plt.show()

File: Preprocessing/CrystalGeneratrion/CreateElectronDist.c
#include <stdio.h>
#include <omp.h>
#include <stdbool.h>
#include <stdlib.h>
#include <complex.h>
#include <math.h>
#include <string.h>
#undef I
complex I = _Complex_I; //Makes it so z = a + b * I

const int nPix = 128; //Size of square
const int FileSize = 841;
const int NumberImages = 19890;
const double RScattFacToVolts = 47.913838;
const double complex constPi = 2.0 * 3.1415926535 * _Complex_I;



int main()
{
  int i, j, k, n;
  double x, y;
  //char Path[] = "/home/ug-ml/Documents/GitHub/FelixML_Project/Preprocessing/CrystalGeneratrion";
  char PathOfPath[] = "/home/ug-ml/felix-ML/DataGenerator3/CifFolder/FilePaths.txt";

  FILE* file_ptr1 = fopen(PathOfPath, "r"); //writing to a file called Frac1
  char*** FilePaths = (char***)calloc(NumberImages, sizeof(char**));
  char As[2000], Bs[2000], Cs[2000];
  int Ad, Bd;
  double Af;
  for(i = 0; i < NumberImages; i++)
  {
    fscanf(file_ptr1, "%s %s %s %d %d %lf", As, Bs, Cs, &Ad, &Bd, &Af);
    FilePaths[i] = (char**)calloc(2, sizeof(char*));
    for(j = 0; j < 2; j++)
    {
      FilePaths[i][j] = (char*)calloc(2000, sizeof(char));
    }
    strcpy(FilePaths[i][0], Bs);
    strcpy(FilePaths[i][1], Cs);
    //printf("%s %s\n", FilePaths[i][0], FilePaths[i][1]);
  }
  fclose(file_ptr1);





  void OpenData(char*, double complex*, int**);

  double ReDotgVec;
  double complex ExpVar;
  double OneOvernPix = 1.0 / (nPix - 1);
  FILE* file_ptr2;



  #pragma omp parallel default(none) private(i, j, k, n, x, y, ReDotgVec, ExpVar, file_ptr2) shared(OneOvernPix, FilePaths)
  {
    double complex* Cug = (double complex*)calloc(FileSize, sizeof(double complex));
    int** gVec = (int**)calloc(FileSize, sizeof(int*));
    for(i = 0; i < FileSize; i++)
    {
      gVec[i] = (int*)calloc(3, sizeof(int));
    }

    double** RreUr = (double**)calloc(nPix, sizeof(double*));
    for(i = 0; i < nPix; i++)
    {
      RreUr[i] = (double*)calloc(nPix, sizeof(double));
    }
    //printf("Hello\n");

    #pragma omp for
    for(k = 0; k < NumberImages; k++)
    {

      //printf("%d %d %d\n", RreUr, gVec, Cug);
      //printf("Reading File\n");
      OpenData(FilePaths[k][1], Cug, gVec);
      //printf("%s\n", FilePaths[k][1]);
      for(i = 0; i < nPix; i++)
      {
        x = i * OneOvernPix;
        for(j = 0; j < nPix; j++)
        {
          y = j * OneOvernPix;
          for(n = 0; n < FileSize; n++)
          {
            //ReDotgVec = np.dot(Re, gVectors[n])
            ReDotgVec = x * gVec[n][0] + y * gVec[n][1];
            ExpVar = cexp(constPi * ReDotgVec);
            RreUr[j][i]+= creal(Cug[n] * ExpVar);
          }

        }
      }
      file_ptr2 = fopen(FilePaths[k][0], "w"); //writing to a file called Frac1
      for(i = 0; i < nPix; i++)
      {
        for(j = 0; j < nPix; j++)
        {
          RreUr[j][i] = RreUr[j][i] * RScattFacToVolts;
          //printf("%g ", RreUr[j][i]);
          fprintf(file_ptr2, "%g ", RreUr[j][i]);
          RreUr[j][i] = 0;
        }
        //printf("\n");
        fprintf(file_ptr2, "\n");
      }
      printf("%d of %d\n", k, NumberImages);
      fclose(file_ptr2);
    }
  }
  return(0);
}



void OpenData(char* NameofFile, double complex* Cug, int** gVec)
{
  FILE * file_ptr = fopen(NameofFile, "r"); //writing to a file called Frac1
  int i;
  double A, B, C, D;
  int E, F, G;
  //printf("%s\n ", NameofFile);
  for(i = 0; i < FileSize; i++)
  {
    fscanf(file_ptr, "\t%d\t%d\t%d\t%lf\t%lf\t%lf\t%lf", &gVec[i][0], &gVec[i][1], &gVec[i][2], &A, &B, &C, &D);
    Cug[i] = C + I * D;
    //printf("%g + i%g\n", creal(Cug[i]), cimag(Cug[i]));
    //printf("%d %d %d\n", gVec[i][0], gVec[i][1], gVec[i][2]);
  }
  fclose(file_ptr);
}




char* concat(char* s1, char* s2)
{
    char *result = malloc(strlen(s1) + strlen(s2) + 1); // +1 for the null-terminator
    // in real code you would check for errors in malloc here
    strcpy(result, s1);
    strcat(result, s2);
    return result;
}
